lib/core/src/Cardano/Wallet/Primitive/Collateral.hs

```diff
@@ -0,0 +1,248 @@
+    BaseAddress CredentialKeyHash CredentialScriptHash    -> 0b0010
+    BaseAddress CredentialScriptHash CredentialScriptHash -> 0b0011
+    PointerAddress CredentialKeyHash                      -> 0b0100
+    PointerAddress CredentialScriptHash                   -> 0b0101
+    EnterpriseAddress CredentialKeyHash                   -> 0b0110
+    EnterpriseAddress CredentialScriptHash                -> 0b0111
+    BootstrapAddress                                      -> 0b1000
+    StakeAddress CredentialKeyHash                        -> 0b1110
+    StakeAddress CredentialScriptHash                     -> 0b1111
+
+-- | Write an AddressType to a binary stream.
+putAddressType :: AddressType -> B.Put
+putAddressType t =
+    B.putWord8 $
+    fromIntegral @Word4 @Word8 (addressTypeToHeaderNibble t) `Bits.shiftL` 4
+
+-- By inspecting the bit pattern of an Address, we can determine it's address
+-- type.
+
+-- | Get the address type of a given address.
+addressType :: Address -> Maybe AddressType
+addressType (Address bytes) =
+    case B.runGetOrFail getAddressType (BL.fromStrict bytes) of
+        Left _ ->
+            Nothing
+        Right (_, _, addrType) ->
+            Just addrType
+
+-- The funds associated with an address are considered suitable for use as
+-- collateral iff the payment credential column of that address is "key hash".
+
+-- | A simple function which determines if an @AddressType@ is suitable for use
+-- as collateral. Only @AddressType@s with a "key hash" payment credential are
+-- considered suitable for use as collateral.
+addressTypeSuitableForCollateral :: AddressType -> Bool
+addressTypeSuitableForCollateral = \case
+    BaseAddress CredentialKeyHash _ ->
+         True
+    BaseAddress CredentialScriptHash _ ->
+         False
+
+    PointerAddress CredentialKeyHash ->
+         True
+    PointerAddress CredentialScriptHash ->
+         False
+
+    EnterpriseAddress CredentialKeyHash ->
+         True
+    EnterpriseAddress CredentialScriptHash ->
+         False
+
+    StakeAddress _ ->
+         False
+
+    BootstrapAddress ->
+         True
+
+-- | Reasons why an address might be considered unsuitable for a collateral
+-- input.
+data AddrNotSuitableForCollateral
+    = IsScriptAddr
+    -- ^ The address is some form of script address
+    | IsStakeAddr
+    -- ^ The address is some form of stake address
+    | IsMalformedOrUnknownAddr
+    -- ^ The address could not be parsed
+    deriving (Eq, Show)
+
+-- | Analyze an address to determine if its funds are suitable for use as a
+-- collateral input. Slight caveat: this function presumes that the given
+-- address is well-formed, it does not validate the address.
+--
+-- This function returns an Either instead of a Maybe so that we can test
+-- that it has failed for the right reason.
+classifyCollateralAddress
+    :: Address
+    -> Either AddrNotSuitableForCollateral Address
+classifyCollateralAddress addr@(Address addrBytes) =
```

Comment:
Hi @sevanspowell! I think there is a good simplification we can make here:

It seems to me that `classifyCollateralAddress` and `addressTypeSuitableForCollateral` are duplicating some of the same logic. (Though please correct me if you think I'm wrong about this, or have missed something!)

Suppose we define the following function:
```hs
addressSuitableForCollateral :: Address -> Bool
addressSuitableForCollateral =
    maybe False addressTypeSuitableForCollateral . addressType
```
Then we can use this function wherever we would have used `classifyCollateralAddress` (more or less).

This also allows us to remove the `classifyCollateralAddress` function and associated `AddrNotSuitableForCollateral` type, with no change in the behaviour, and no loss of functionality.

I've coded up this suggestion in a PR here: https://github.com/input-output-hk/cardano-wallet/pull/2814

What do you think?

Reply:
Yeah, I think that's good. The only reason I kept it was so that debugging is a bit easier (it tells you why, rather than just "no"). But it's not a super important point, I like your PR :+1: